src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -436,15 +437,29 @@ private Type GetBaseTypes(Parser parser, TypeDefinitionAst typeDefinitionAst, ou
 
             private bool ShouldImplementProperty(string name, Type type)
             {
-                foreach (var interfaceType in _typeBuilder.GetInterfaces())
+                if (_interfaceProperties == null)
                 {
-                    if (interfaceType.GetProperty(name, type) != null)
+                    _interfaceProperties = new HashSet<Tuple<string, Type>>();
+                    foreach (var interfaceType in _typeBuilder.GetInterfaces())
                     {
-                        return true;
+                        // TypeBuilder.GetInterfaces() returns only the interfaces that was explicitly passed to its constructor.
+                        // During compilation the interface hierarchy is flattened, so we only need to resolve one level of ancestral interfaces.
+                        foreach (var parentInterface in interfaceType.GetInterfaces())
+                        {
+                            foreach (var property in parentInterface.GetProperties())
+                            {
+                                _interfaceProperties.Add(new Tuple<string, Type>(property.Name, property.PropertyType));
+                            }
```

Comment:
Can you please put this code (449 - 452) in a local function? This code is used twice here, and it will be cleaner to wrap it in a local function.

Reply:
A separate function for a single loop seems a bit overkill, I was thinking about separating interface resolution and property enumeration instead, a la:

```csharp
_interfaceProperties = new HashSet<Tuple<string, Type>>();

// TypeBuilder.GetInterfaces() returns only the interfaces that was explicitly passed to its constructor.
// During compilation the interface hierarchy is flattened, so we only need to resolve one level of ancestral interfaces.
var allInterfaces = new HashSet<Type>();
foreach (var interfaceType in _typeBuilder.GetInterfaces())
{
    foreach (var parentInterface in interfaceType.GetInterfaces())
    {
        allInterfaces.Add(parentInterface);
    }
    allInterfaces.Add(interfaceType);
}

foreach (var interfaceType in allInterfaces)
{
    foreach (var property in interfaceType.GetProperties())
    {
        _interfaceProperties.Add(Tuple.Create(property.Name, property.PropertyType));
    }
}
```

Reply:
This looks great :)